### PR TITLE
 New db structure and trigger reorg from orchestrator

### PR DESCRIPTION
**Pull request by atif-konasl**
## Motivation
Removing slot key and introducing stepId concept because slot can be missed and it does not consecutive. As well as orchestrator should support multi shard verification, so new db structure is needed.

### CODE REVIEW

**Comment by blazejkrzak** [comment]
tests are failing

**Review by blazejkrzak — CHANGES_REQUESTED** [review]
tests are not passing

File: orchestrator/consensus/handler.go [hunk 1]
```diff
@@ -1,111 +1,179 @@
...
+
+	return s.insertIntoChain(vanShardInfo, headerInfo.Header)
 }
 
 // processVanguardShardInfo
 func (s *Service) processVanguardShardInfo(vanShardInfo *types.VanguardShardInfo) error {
 	slot := vanShardInfo.Slot
+
+	// short circuit check, if this header is already in verified sharding info db then send confirmation instantly
+	if shardInfo := s.getShardingInfo(slot); shardInfo != nil {
```
Comment (blazejkrzak) [inline orchestrator/consensus/handler.go:46]: you can reverse logic to get rid of if nesting

File: orchestrator/rpc/api/events/event_api.go [hunk 2]
```diff
@@ -8,6 +8,110 @@ import (
 	"github.com/pkg/errors"
 )
 
+// MinimalConsensusInfo
+func (api *PublicFilterAPI) MinimalConsensusInfo(ctx context.Context, requestedEpoch uint64) (*rpc.Subscription, error) {
```
Comment (blazejkrzak) [inline orchestrator/rpc/api/events/event_api.go:12]: github can be misleading me but is it totally new endpoint?

File: orchestrator/vanguardchain/setup.go [hunk 3]
```diff
@@ -1,77 +1,61 @@
...
-		totalConsensusInfos = append(totalConsensusInfos, epochInfoV2)
-		err = vanguardDb.SaveConsensusInfo(ctx, epochInfoV2)
-		assert.NoError(t, err)
-	}
-	err = vanguardDb.SaveLatestEpoch(ctx, uint64(numberOfElements-1))
-	assert.NoError(t, err)
-
-	return vanguardDb
-}
+//func serviceInit(t *testing.T, numberOfElements byte) (*Service, *logTest.Hook) {
```
Comment (blazejkrzak) [inline orchestrator/vanguardchain/setup.go:3]: this is bad

**Review by blazejkrzak — CHANGES_REQUESTED** [review]
- uncomment tests or provide alternative
- functions that are part of interface must be tested
- nesting of ifs and else statements can be avoided with little effort

File: orchestrator/consensus/handler.go [hunk 4]
```diff
@@ -1,111 +1,179 @@
...
-	"fmt"
+	"github.com/lukso-network/lukso-orchestrator/orchestrator/utils"
 
 	"github.com/ethereum/go-ethereum/common"
 	eth1Types "github.com/ethereum/go-ethereum/core/types"
 	"github.com/lukso-network/lukso-orchestrator/shared/types"
 )
 
-// processPandoraHeader
+// processPandoraHeader method process incoming pandora shard header from pandora chain
```
Comment (VOID404) [inline orchestrator/consensus/handler.go:11]: The first sentence of a comment should be a one-sentence summary that starts with the name being declared.
```suggestion
// processPandoraHeader processes pandora shard header coming from pandora chain
```

File: orchestrator/consensus/handler.go [hunk 5]
```diff
@@ -1,111 +1,179 @@
...
+				PandoraHeaderHash: shardInfo.Shards[0].Blocks[0].HeaderRoot,
+				Status:            types.Verified,
+			})
+			return nil
+		}
+	}
+
+	// retrieving sharding
 	s.pandoraPendingHeaderCache.Put(s.ctx, slot, headerInfo.Header)
 	vanShardInfo, _ := s.vanguardPendingShardingCache.Get(s.ctx, slot)
```
Comment (VOID404) [inline orchestrator/consensus/handler.go:33]: Error handling?

File: orchestrator/consensus/handler.go [hunk 6]
```diff
@@ -1,111 +1,179 @@
...
+				VanguardBlockHash: shardInfo.SlotInfo.BlockRoot,
+				PandoraHeaderHash: shardInfo.Shards[0].Blocks[0].HeaderRoot,
+				Status:            types.Verified,
+			})
+			return nil
+		}
+	}
+
+	// retrieving sharding
 	s.pandoraPendingHeaderCache.Put(s.ctx, slot, headerInfo.Header)
```
Comment (VOID404) [inline orchestrator/consensus/handler.go:32]: Error handling?

File: orchestrator/consensus/handler.go [hunk 7]
```diff
@@ -1,111 +1,179 @@
...
+	return s.insertIntoChain(vanShardInfo, header)
 }
 
-// verifyShardingInfo
-func (s *Service) verifyShardingInfo(slot uint64, vanShardInfo *types.VanguardShardInfo, header *eth1Types.Header) error {
-	slotInfo := &types.SlotInfo{
-		PandoraHeaderHash: header.Hash(),
-		VanguardBlockHash: common.BytesToHash(vanShardInfo.BlockHash[:]),
+// insertIntoChain method
+//	- verifies shard info and pandora header
```
Comment (VOID404) [inline orchestrator/consensus/handler.go:63]: In this comment block, this is the only line with a tab

File: orchestrator/consensus/handler.go [hunk 8]
```diff
@@ -1,111 +1,179 @@
...
+//  - write into db
```
Comment (VOID404) [inline orchestrator/consensus/handler.go:64]: Either 3rd person everywhere or imperative everywhere

File: orchestrator/consensus/test_setup.go [hunk 9]
```diff
@@ -22,20 +23,12 @@ func (mc *mockFeedService) SubscribeShutdownSignalEvent(signals chan<- *types.Re
 	return mc.scope.Track(mc.subscriptionShutdownFeed.Subscribe(signals))
 }
 
-func (mc *mockFeedService) ReSubscribeBlocksEvent() error {
+func (mc *mockFeedService) StopSubscription(reorgInfo *types.Reorg) {
 	panic("implement me")
```
Comment (mxmar) [inline orchestrator/consensus/test_setup.go:27]: I think we can improve this - provide proper tests. Me or Wojtek (@VOID404) can take care of this tomorrow.

**Review by mxmar — CHANGES_REQUESTED** [review]
I've left some comments. I think we need to solve tests first and then comments. Probably test is failing because of:
```
# github.com/lukso-network/lukso-orchestrator/orchestrator/vanguardchain [github.com/lukso-network/lukso-orchestrator/orchestrator/vanguardchain.test]
orchestrator/vanguardchain/integration_test.go:14:11: undefined: dbSetup
```

File: orchestrator/consensus/handler.go [hunk 10]
```diff
@@ -1,111 +1,179 @@
...
 }
 
-func (s *Service) reorgDB(revertSlot uint64) error {
+func (s *Service) getShardingInfo(slot uint64) *types.MultiShardInfo {
 	// Removing slot infos from verified slot info db
-	if err := s.verifiedSlotInfoDB.RemoveRangeVerifiedInfo(revertSlot+1, s.verifiedSlotInfoDB.LatestSavedVerifiedSlot()); err != nil {
-		log.WithError(err).Error("found error while reverting orchestrator database in reorg phase")
+	stepId, err := s.db.GetStepIdBySlot(slot)
+	if err != nil {
+		log.WithError(err).WithField("slot", slot).Error("Could not found step id from DB")
```
Comment (VOID404) [inline orchestrator/consensus/handler.go:124]:
```suggestion
		log.WithError(err).WithField("slot", slot).Error("Could not find step id from DB")
```

File: orchestrator/consensus/handler.go [hunk 11]
```diff
@@ -1,111 +1,179 @@
...
+		return nil
+	}
+
+	shardInfo, err := s.db.VerifiedShardInfo(stepId)
+	if err != nil {
+		log.WithError(err).WithField("stepId", stepId).Error("Could not found shard info from DB during reorg")
```
Comment (VOID404) [inline orchestrator/consensus/handler.go:130]:
```suggestion
		log.WithError(err).WithField("stepId", stepId).Error("Could not find shard info from DB during reorg")
```

File: orchestrator/consensus/handler.go [hunk 12]
```diff
@@ -1,111 +1,179 @@
...
+		return nil
+	}
+
+	if shardInfo == nil {
+		log.WithField("stepId", stepId).Debug("Could not found shard info from DB during reorg")
```
Comment (VOID404) [inline orchestrator/consensus/handler.go:135]:
```suggestion
		log.WithField("stepId", stepId).Debug("Could not find shard info from DB during reorg")
```

File: orchestrator/consensus/handler.go (hunk 1 above)
Comment (VOID404) [reply orchestrator/consensus/handler.go:46]: Logic operators would solve this issue too

File: orchestrator/consensus/handler.go [hunk 13]
```diff
@@ -1,111 +1,179 @@
...
+func (s *Service) writeFinalizeInfo(finalizeSlot, finalizeEpoch uint64) {
+	curFinalizeSlot := s.db.FinalizedSlot()
+	if finalizeSlot > curFinalizeSlot {
+		if err := s.db.SaveFinalizedSlot(finalizeSlot); err != nil {
+			log.WithError(err).Warn("Failed to store new finalized info")
+		}
+	}
+
+	curFinalizeEpoch := s.db.FinalizedEpoch()
+	if finalizeEpoch > curFinalizeEpoch {
```
[the remark this answers is not in the document]
Comment (VOID404) [reply orchestrator/consensus/handler.go:174]: Logic operators would solve this issue too

File: orchestrator/consensus/reorg.go [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/lukso-network/lukso-orchestrator/shared/types"
+	"github.com/pkg/errors"
+	"sync/atomic"
+)
+
+// processReorg
+func (s *Service) processReorg(parentVanBlkHash common.Hash) error {
```
[the remark this answers is not in the document]
Comment (VOID404) [reply orchestrator/consensus/reorg.go:11]: Also, if we need a function to be testable but don't want to expose it, we can create an internal package. This way it can be "public", but only to our package.

File: orchestrator/consensus/reorg.go [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+	atomic.StoreUint32(&s.reorgInProgress, 1)
+	defer func() {
```
Comment (VOID404) [inline orchestrator/consensus/reorg.go:13]: This anonymous function seems pointless

File: orchestrator/consensus/reorg.go [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+		atomic.StoreUint32(&s.reorgInProgress, 0)
+	}()
+
+	finalizedSlot := s.db.FinalizedSlot()
+	finalizedStepId, err := s.db.GetStepIdBySlot(finalizedSlot)
+	if err != nil {
+		log.WithError(err).WithField("finalizedSlot", finalizedSlot).WithField("latestFinalizedStepId", finalizedStepId).
+			Error("Could not found step id from DB")
```
Comment (VOID404) [inline orchestrator/consensus/reorg.go:21]:
```suggestion
			Error("Could not find step id from DB")
```

File: orchestrator/consensus/reorg.go [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+		return err
+	}
+
+	latestStepId := s.db.LatestStepID()
+	parentShardInfo, err := s.db.FindAncestor(latestStepId, finalizedStepId, parentVanBlkHash)
+	if err != nil || parentShardInfo == nil {
+		log.WithField("finalizedSlot", finalizedSlot).WithField("latestFinalizedStepId", finalizedStepId).
+			Error("Could not found parent shard info from DB")
```
Comment (VOID404) [inline orchestrator/consensus/reorg.go:29]:
```suggestion
			Error("Could not find parent shard info from DB")
```

File: orchestrator/db/kv/kv.go [hunk 18]
```diff
@@ -78,7 +76,17 @@ func NewKVStore(ctx context.Context, dirPath string, config *Config) (*Store, er
 	if err != nil {
 		return nil, err
 	}
-	verifiedSlotInfoCache, err := ristretto.NewCache(&ristretto.Config{
+
+	multiShardsInfoCache, err := ristretto.NewCache(&ristretto.Config{
+		NumCounters: 1000,                    // number of keys to track frequency of (1000).
```
Comment (VOID404) [inline orchestrator/db/kv/kv.go:81]: Hardcoded constants should be extracted into a `const`

File: orchestrator/db/kv/multi_shard_info.go [hunk 19]
```diff
@@ -0,0 +1,277 @@
...
+		value := bkt.Get(key[:])
+		if value == nil {
+			return nil
+		}
+		return decode(value, &shardInfo)
+	})
+	return shardInfo, err
+}
+
+// ConsensusInfos
```
Comment (VOID404) [inline orchestrator/db/kv/multi_shard_info.go:29]: Documentation comment should start with the name of the function

File: orchestrator/db/kv/multi_shard_info.go [hunk 20]
```diff
@@ -0,0 +1,277 @@
...
+			if err != nil {
+				log.WithField("step", step).Warn("DB is corrupted")
+				return err
+			}
+			shards[step] = shardInfo
+		}
+		return nil
+	})
+	// the query not successful
+	if err != nil {
```
Comment (VOID404) [inline orchestrator/db/kv/multi_shard_info.go:50]: This `if` is pointless.

File: orchestrator/db/kv/multi_shard_info.go [hunk 21]
```diff
@@ -0,0 +1,277 @@
...
+	})
+}
+
+// SaveLatestStepID
+func (s *Store) SaveLatestStepID(stepID uint64) error {
+	// storing latest epoch number into db
+	return s.db.Update(func(tx *bolt.Tx) error {
+		bkt := tx.Bucket(latestInfoMarkerBucket)
+		stepBytes := bytesutil.Uint64ToBytesBigEndian(stepID)
+		if err := bkt.Put(latestStepIdKey, stepBytes); err != nil {
```
Comment (VOID404) [inline orchestrator/db/kv/multi_shard_info.go:85]: This if is pointless

File: orchestrator/db/kv/multi_shard_info.go [hunk 22]
```diff
@@ -0,0 +1,277 @@
...
+		return nil
+	})
+}
+
+// LatestStepID
+func (s *Store) LatestStepID() uint64 {
+	var latestStepId uint64
+	// Db is not prepared yet. Retrieve latest saved epoch number from db
+	if !s.isRunning {
+		s.db.View(func(tx *bolt.Tx) error {
```
Comment (VOID404) [inline orchestrator/db/kv/multi_shard_info.go:97]: unhandled error

File: orchestrator/db/kv/multi_shard_info.go [hunk 23]
```diff
@@ -0,0 +1,277 @@
...
+		// not found the latest epoch in db. so latest epoch will be zero
+		if stepIdBytes == nil {
+			stepId = uint64(0)
+			return nil
+		}
+		stepId = bytesutil.BytesToUint64BigEndian(stepIdBytes)
+		return nil
+	})
+
+	if err != nil {
```
Comment (VOID404) [inline orchestrator/db/kv/multi_shard_info.go:174]: this if is pointless

File: orchestrator/db/kv/multi_shard_info.go [hunk 24]
```diff
@@ -0,0 +1,277 @@
...
+		return nil
+	})
+}
+
+// LatestLatestFinalizedSlot
+func (s *Store) FinalizedSlot() uint64 {
+	var latestFinalizedSlot uint64
+	// Db is not prepared yet. Retrieve latest saved finalized slot number from db
+	if !s.isRunning {
+		s.db.View(func(tx *bolt.Tx) error {
```
Comment (VOID404) [inline orchestrator/db/kv/multi_shard_info.go:199]: unhandled error

File: orchestrator/db/kv/multi_shard_info.go [hunk 25]
```diff
@@ -0,0 +1,277 @@
...
+			bkt := tx.Bucket(latestInfoMarkerBucket)
+			slotBytes := bkt.Get(latestFinalizedSlotKey[:])
+			// not found the latest finalized slot in db. so latest finalized slot will be zero
+			if slotBytes == nil {
+				latestFinalizedSlot = uint64(0)
+				log.Trace("Latest finalized slot number could not find in db. It may happen for brand new DB")
```
Comment (VOID404) [inline orchestrator/db/kv/multi_shard_info.go:205]: The slot number could not find _what_ in DB?

Comment (VOID404) [inline orchestrator/db/kv/multi_shard_info.go:205]:
```suggestion
				log.Trace("Latest finalized slot number could not find in DB. It may happen for brand new DB")
```

File: orchestrator/db/kv/multi_shard_info.go [hunk 26]
```diff
@@ -0,0 +1,277 @@
...
+	var latestFinalizedEpoch uint64
+	// Db is not prepared yet. Retrieve latest saved finalized slot number from db
+	if !s.isRunning {
+		s.db.View(func(tx *bolt.Tx) error {
+			bkt := tx.Bucket(latestInfoMarkerBucket)
+			epochBytes := bkt.Get(latestFinalizedEpochKey[:])
+			// not found the latest finalized slot in db. so latest finalized slot will be zero
+			if epochBytes == nil {
+				latestFinalizedEpoch = uint64(0)
+				log.Trace("Latest finalized epoch number could not find in db. It may happen for brand new DB")
```
Comment (VOID404) [inline orchestrator/db/kv/multi_shard_info.go:240]: The epoch number could not find _what_ in DB?

Comment (VOID404) [inline orchestrator/db/kv/multi_shard_info.go:240]:
```suggestion
				log.Trace("Latest finalized epoch number could not find in DB. It may happen for brand new DB")
```

File: orchestrator/rpc/api/api_backend.go [hunk 27]
```diff
@@ -54,75 +55,112 @@ func (backend *Backend) ConsensusInfoByEpochRange(fromEpoch uint64) ([]*types.Mi
...
+func (b *Backend) VerifiedSlotInfos(fromSlot uint64) (map[uint64]*types.BlockStatus, error) {
+	// Short circuit for slot zero
+	if fromSlot == 0 {
+		return nil, nil
+	}
+
+	// Removing slot infos from verified slot info db
+	stepId, err := b.verifiedShardInfoDB.GetStepIdBySlot(fromSlot)
+	if err != nil {
+		log.WithError(err).WithField("fromSlot", fromSlot).WithField("stepId", stepId).Error("Could not found step id from DB")
```
Comment (VOID404) [inline orchestrator/rpc/api/api_backend.go:79]:
```suggestion
		log.WithError(err).WithField("fromSlot", fromSlot).WithField("stepId", stepId).Error("Could not find step id from DB")
```

File: orchestrator/rpc/api/api_backend.go [hunk 28]
```diff
@@ -54,75 +55,112 @@ func (backend *Backend) ConsensusInfoByEpochRange(fromEpoch uint64) ([]*types.Mi
...
+		return nil, errors.Wrap(err, "Could not found step id from DB")
```
Comment (VOID404) [inline orchestrator/rpc/api/api_backend.go:80]:
```suggestion
		return nil, errors.Wrap(err, "Could not find step id from DB")
```

File: orchestrator/rpc/api/api_backend.go [hunk 29]
```diff
@@ -54,75 +55,112 @@ func (backend *Backend) ConsensusInfoByEpochRange(fromEpoch uint64) ([]*types.Mi
...
+	}
+
+	shardInfos, err := b.verifiedShardInfoDB.VerifiedShardInfos(stepId)
+	if err != nil {
+		return nil, errors.Wrap(err, "Could not found verified shard infos from DB")
```
Comment (VOID404) [inline orchestrator/rpc/api/api_backend.go:85]:
```suggestion
		return nil, errors.Wrap(err, "Could not find verified shard infos from DB")
```

File: orchestrator/rpc/api/api_backend.go [hunk 30]
```diff
@@ -54,75 +55,112 @@ func (backend *Backend) ConsensusInfoByEpochRange(fromEpoch uint64) ([]*types.Mi
...
-		if !requestFrom && vanHeaderHash != hash {
-			log.WithError(ErrHeaderHashMisMatch).
-				Warn("Failed to match header hash with requested header hash from vanguard node")
-			logPrinter(types.Invalid)
-			return types.Invalid
-		}
+	// Removing slot infos from verified slot info db
+	stepId, err := b.verifiedShardInfoDB.GetStepIdBySlot(slot)
+	if err != nil {
+		log.WithError(err).WithField("slot", slot).WithField("stepId", stepId).Error("Could not found step id from DB")
```
Comment (VOID404) [inline orchestrator/rpc/api/api_backend.go:132]:
```suggestion
		log.WithError(err).WithField("slot", slot).WithField("stepId", stepId).Error("Could not find step ID from DB")
```

File: orchestrator/rpc/api/api_backend.go [hunk 31]
```diff
@@ -54,75 +55,112 @@ func (backend *Backend) ConsensusInfoByEpochRange(fromEpoch uint64) ([]*types.Mi
...
+		return types.Invalid
+	}
 
-		status = types.Verified
-		logPrinter(types.Verified)
-		return status
+	// finally found in the database so return immediately so that no other db call happens
+	if shardInfo, _ := b.verifiedShardInfoDB.VerifiedShardInfo(stepId); shardInfo != nil {
+		if len(shardInfo.Shards) > 0 && len(shardInfo.Shards[0].Blocks) > 0 {
```
[the remark this answers is not in the document]
Comment (VOID404) [reply orchestrator/rpc/api/api_backend.go:138]: Logic operators would work too

File: orchestrator/rpc/api/api_backend.go [hunk 32]
```diff
@@ -54,75 +55,112 @@ func (backend *Backend) ConsensusInfoByEpochRange(fromEpoch uint64) ([]*types.Mi
...
+	if err != nil {
+		log.WithError(err).WithField("slot", slot).WithField("stepId", stepId).Error("Could not found step id from DB")
+		return types.Invalid
+	}
 
-		status = types.Verified
-		logPrinter(types.Verified)
-		return status
+	// finally found in the database so return immediately so that no other db call happens
+	if shardInfo, _ := b.verifiedShardInfoDB.VerifiedShardInfo(stepId); shardInfo != nil {
```
Comment (VOID404) [inline orchestrator/rpc/api/api_backend.go:137]: Unhandled error

File: orchestrator/rpc/api/api_backend.go [hunk 33]
```diff
@@ -54,75 +55,112 @@ func (backend *Backend) ConsensusInfoByEpochRange(fromEpoch uint64) ([]*types.Mi
...
 func (backend *Backend) LatestFinalizedSlot() uint64 {
-	return backend.VerifiedSlotInfoDB.LatestLatestFinalizedSlot()
+	return backend.verifiedShardInfoDB.FinalizedSlot()
 }
 
 // GetSlotStatus
-func (backend *Backend) GetSlotStatus(ctx context.Context, slot uint64, hash common.Hash, requestFrom bool) types.Status {
+func (b *Backend) GetSlotStatus(ctx context.Context, slot uint64, hash common.Hash, requestFrom bool) types.Status {
 	// by default if nothing is found then return skipped
 	status := types.Pending
```
Comment (VOID404) [inline orchestrator/rpc/api/api_backend.go:121]: This variable is pointless

File: orchestrator/rpc/api/backend_service_mock.go [hunk 34]
```diff
@@ -1,16 +1,17 @@
...
 	"context"
 	"github.com/ethereum/go-ethereum/common"
 	eth1Types "github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/event"
+	"github.com/lukso-network/lukso-orchestrator/orchestrator/rpc/api/events"
 	eventTypes "github.com/lukso-network/lukso-orchestrator/shared/types"
 	"time"
 )
 
 var (
```
Comment (VOID404) [inline orchestrator/rpc/api/backend_service_mock.go:13]:
```suggestion
const (
```

File: orchestrator/rpc/api/events/event_api.go [hunk 35]
```diff
@@ -8,6 +8,110 @@ import (
...
+	notifier, supported := rpc.NotifierFromContext(ctx)
+	if !supported {
+		return &rpc.Subscription{}, rpc.ErrNotificationsUnsupported
```
Comment (VOID404) [inline orchestrator/rpc/api/events/event_api.go:15]:
```suggestion
		return nil, rpc.ErrNotificationsUnsupported
```

File: orchestrator/rpc/api/events/event_api.go (hunk 2 above)
Comment (VOID404) [inline orchestrator/rpc/api/events/event_api.go:12]: This function should be split into some smaller functions, it's cyclomatic complexity is very high, so it's very hard to follow what it does.

File: orchestrator/rpc/api/events/event_api.go [hunk 36]
```diff
@@ -8,6 +8,110 @@ import (
...
+	}
+	rpcSub := notifier.CreateSubscription()
+
+	go func() {
+
+		batchSender := func(start, end uint64) error {
+			epochInfos, err := api.backend.ConsensusInfoByEpochRange(start)
+			if err != nil {
+				log.WithError(err).Error("Some epoch infos are missing in db.")
```
Comment (VOID404) [inline orchestrator/rpc/api/events/event_api.go:24]:
```suggestion
				log.WithError(err).Error("Some epoch info is missing in db.")
```

File: orchestrator/rpc/api/events/event_api.go [hunk 37]
```diff
@@ -8,6 +8,110 @@ import (
...
+				return errors.Wrap(err, "Missing epoch infos in db. Could not send over stream.")
```
Comment (VOID404) [inline orchestrator/rpc/api/events/event_api.go:25]:
```suggestion
				return errors.Wrap(err, "Missing epoch info in DB. Could not send over stream.")
```

File: orchestrator/rpc/api/events/event_api.go [hunk 38]
```diff
@@ -8,6 +8,110 @@ import (
...
+					Epoch:            ei.Epoch,
+					ValidatorList:    ei.ValidatorList,
+					EpochStartTime:   ei.EpochStartTime,
+					SlotTimeDuration: ei.SlotTimeDuration,
+					FinalizedSlot:    latestFinalizedSlot,
+				}); err != nil {
+					log.WithField("start", start).
+						WithField("end", end).
+						WithError(err).
+						Error("Failed to send epoch info. Could not send over stream.")
```
Comment (VOID404) [inline orchestrator/rpc/api/events/event_api.go:39]: Inconsistent wording in logs. "Failed" -> "Could not"
```suggestion
						Error("Could not send epoch info over stream.")
```

File: orchestrator/rpc/api/events/event_api.go [hunk 39]
```diff
@@ -8,6 +8,110 @@ import (
...
+					return errors.Wrap(err, "Failed to send epoch info. Could not send over stream.")
+				}
+				log.WithField("epoch", ei.Epoch).WithField("latestFinalizedSlot", latestFinalizedSlot).
+					Info("published epoch info to pandora")
```
Comment (VOID404) [inline orchestrator/rpc/api/events/event_api.go:43]: Inconsistent formatting

```suggestion
					Info("Published epoch info to pandora")
```

File: orchestrator/rpc/api/events/event_api.go [hunk 40]
```diff
@@ -8,6 +8,110 @@ import (
...
+					Info("Sending consensus info to subscriber")
+
+				if firstTime {
+					firstTime = false
+					startEpoch = endEpoch
+					endEpoch = api.backend.LatestEpoch()
+
+					if startEpoch+1 < endEpoch {
+						log.WithField("startEpoch", startEpoch).WithField("endEpoch", endEpoch).
+							Debug("successfully published left over epoch infos")
```
Comment (VOID404) [inline orchestrator/rpc/api/events/event_api.go:75]: Inconsistent log formatting
```suggestion
							Debug("Successfully published left over epoch infos")
```

File: orchestrator/rpc/api/events/event_api.go [hunk 41]
```diff
@@ -8,6 +8,110 @@ import (
...
+						if err := batchSender(startEpoch, endEpoch); err != nil {
+							return
+						}
+					}
+					log.WithField("liveSyncEpoch", endEpoch+1).Debug("start publishing live epoch info to pandora")
```
Comment (VOID404) [inline orchestrator/rpc/api/events/event_api.go:80]: Inconsistent log formatting
```suggestion
					log.WithField("liveSyncEpoch", endEpoch+1).Debug("Start publishing live epoch info to pandora")
```

File: orchestrator/rpc/api/events/event_api.go [hunk 42]
```diff
@@ -8,6 +8,110 @@ import (
...
+					Epoch:            currentEpochInfo.Epoch,
+					ValidatorList:    currentEpochInfo.ValidatorList,
+					EpochStartTime:   currentEpochInfo.EpochStartTime,
+					SlotTimeDuration: currentEpochInfo.SlotTimeDuration,
+					ReorgInfo:        currentEpochInfo.ReorgInfo,
+					FinalizedSlot:    currentEpochInfo.FinalizedSlot,
+				})
+				if nil != err {
+					log.WithField("epoch", currentEpochInfo.Epoch).WithError(err).Error(
+						"Failed to notify consensus info")
```
Comment (VOID404) [inline orchestrator/rpc/api/events/event_api.go:93]: Inconsistent log wording
```suggestion
						"Could not notify consensus info")
```

File: orchestrator/rpc/api/events/event_api.go [hunk 43]
```diff
@@ -8,6 +8,110 @@ import (
...
+					return
+				}
+
+				log.WithField("epoch", currentEpochInfo.Epoch).WithField("latestFinalizedSlot", currentEpochInfo.FinalizedSlot).
+					Info("published epoch info to pandora")
```
Comment (VOID404) [inline orchestrator/rpc/api/events/event_api.go:98]: Inconsistent log formatting
```suggestion
					Info("Published epoch info to pandora")
```

File: orchestrator/rpc/api/events/event_api.go [hunk 44]
```diff
@@ -8,6 +8,110 @@ import (
...
+
+				err := notifier.Notify(rpcSub.ID, &generalTypes.MinimalEpochConsensusInfoV2{
+					Epoch:            currentEpochInfo.Epoch,
+					ValidatorList:    currentEpochInfo.ValidatorList,
+					EpochStartTime:   currentEpochInfo.EpochStartTime,
+					SlotTimeDuration: currentEpochInfo.SlotTimeDuration,
+					ReorgInfo:        currentEpochInfo.ReorgInfo,
+					FinalizedSlot:    currentEpochInfo.FinalizedSlot,
+				})
+				if nil != err {
```
Comment (VOID404) [inline orchestrator/rpc/api/events/event_api.go:91]: Inconsistent `if` order. This order is used 3 times in this file, the one I am suggesting is used 97 times
```suggestion
				if err != nil {
```

File: orchestrator/rpc/api/events/event_api.go [hunk 45]
```diff
@@ -85,8 +185,7 @@ func (api *PublicFilterAPI) SteamConfirmedPanBlockHashes(
 					Status:        slotInfoWithStatus.Status,
 					FinalizedSlot: api.backend.LatestFinalizedSlot(),
 				}); err != nil {
-					log.WithField("hash", slotInfoWithStatus.PandoraHeaderHash).
-						Error("Failed to notify slot info status. Could not send over stream.")
+					log.WithField("hash", slotInfoWithStatus.PandoraHeaderHash).Error("Failed to notify slot info status. Could not send over stream.")
```
Comment (VOID404) [inline orchestrator/rpc/api/events/event_api.go:188]: Why unsplit this line? It is 152 characters wide now

File: orchestrator/utils/multi_shard.go [hunk 46]
```diff
@@ -0,0 +1,65 @@
+package utils
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+	eth1Types "github.com/ethereum/go-ethereum/core/types"
+	"github.com/lukso-network/lukso-orchestrator/shared/types"
+)
+
+func PrepareMultiShardData(
+	vanShardInfo *types.VanguardShardInfo,
```
Comment (VOID404) [inline orchestrator/utils/multi_shard.go:10]: Is there a reason to make those arguments pointers? We don't modify them in this function, are they big enough to worry about copying? Using pointers disables some optimisations and is harder on GC

File: orchestrator/utils/multi_shard.go [hunk 47]
```diff
@@ -0,0 +1,65 @@
...
+		BlockRoot: common.BytesToHash(vanShardInfo.BlockHash),
+	}
+	shardInfo.Shards = shards
+
+	return shardInfo
+}
+
+// ConvertShardInfoToBlockStatus converts sharding info to blockStatus
+func ConvertShardInfoToBlockStatus(
+	shardInfo *types.MultiShardInfo,
```
Comment (VOID404) [inline orchestrator/utils/multi_shard.go:44]: Is there a reason to make those arguments pointers? We don't modify them in this function, are they big enough to worry about copying? Using pointers disables some optimisations and is harder on GC

Without pointers, check for nil can be removed

Comment (VOID404) [reply orchestrator/utils/multi_shard.go:44]: > Why should I pass and do copy the whole object to a function just to get a status
1. Memory safety
2. Optimisation (it's placed in different place in memory, it has to be cleaned in a different way etc)

File: orchestrator/rpc/api/events/event_api.go (hunk 45 above)
Comment (VOID404) [reply orchestrator/rpc/api/events/event_api.go:188]: Are you sure? When I tried `gci`, `goimports` and `go fmt` all allow for a split there

File: orchestrator/rpc/api/events/event_api.go (hunk 2 above)
Comment (VOID404) [reply orchestrator/rpc/api/events/event_api.go:12]: We shouldn't write code that needs refactoring. I don't think it'll help us adding more code that'll need cleaning up. If this functionality is needed quickly as an experiment or something, then maybe we should keep it in a separate branch and not merge it to devel? Technical debt is not a good thing.

File: orchestrator/db/kv/multi_shard_info.go (hunk 23 above)
Comment (VOID404) [reply orchestrator/db/kv/multi_shard_info.go:174]:
```go
if err != nil {
    return stepId, err
}
return stepId, nil
```
works exactly in the same way as
```go
return stepId, err
```
It always returns `stepId` (just like original), and it either returns `nil` or `err` (just like original). 

Propagating errors is good, that `if` doesn't add anything to the logic
